db/database.go

```diff
@@ -1069,129 +1067,155 @@ func (db *Database) UpdateAllDocChannels() (int, error) {
 	changeCount := 0
 	docCount := 0
 
-	var importRow QueryIdRow
-	for results.Next(&importRow) {
-		docid := importRow.Id
-		key := realDocID(docid)
-
-		docCount++
-		documentUpdateFunc := func(doc *Document) (updatedDoc *Document, shouldUpdate bool, updatedExpiry *uint32, err error) {
-			imported := false
-			if !doc.HasValidSyncData() {
-				// This is a document not known to the sync gateway. Ignore it:
-				return nil, false, nil, base.ErrUpdateCancel
-			} else {
-				base.Debugf(base.KeyCRUD, "\tRe-syncing document %q", base.UD(docid))
-			}
+	queryLimit := int(db.Options.ResyncQueryLimit)
+	startSeq := uint64(0)
+	endSeq, err := db.sequences.lastSequence()
+	if err != nil {
+		return 0, err
+	}
 
-			// Run the sync fn over each current/leaf revision, in case there are conflicts:
-			changed := 0
-			doc.History.forEachLeaf(func(rev *RevInfo) {
-				bodyBytes, _, err := db.get1xRevFromDoc(doc, rev.ID, false)
-				if err != nil {
-					base.Warnf("Error getting rev from doc %s/%s %s", base.UD(docid), rev.ID, err)
-				}
-				var body Body
-				if err := body.Unmarshal(bodyBytes); err != nil {
-					base.Warnf("Error unmarshalling body %s/%s for sync function %s", base.UD(docid), rev.ID, err)
-					return
-				}
-				channels, access, roles, syncExpiry, _, err := db.getChannelsAndAccess(doc, body, rev.ID)
-				if err != nil {
-					// Probably the validator rejected the doc
-					base.Warnf("Error calling sync() on doc %q: %v", base.UD(docid), err)
-					access = nil
-					channels = nil
+outerLoop:
+	for {
+		results, err := db.QueryResync(queryLimit, startSeq, endSeq)
+		if err != nil {
+			return 0, err
+		}
+
+		queryRowCount := 0
+		highSeq := uint64(0)
+
+		var importRow QueryIdRow
+		for results.Next(&importRow) {
+			if db.ResyncTerminator.CompareAndSwap(true, false) {
+				break outerLoop
```

Comment:
A break here looks like it's going to fall through to line 1219 and log "finished running...", which would give the impression resync completed.  

Maybe the intention was to fall through and ensure that 'invalidate channel caches of principals' still runs, but I don't know if that's necessary.  If resync is half finished the system is already going to be in an inconsistent state.  Blocking until users get invalidated could be a long process (if there are a very large number of users), it may be reasonable to just return here.

Also, I think this should do results.Close before exiting the loop.